### DISCUSSION TITLE
Adding Name to UI things and allowing retrieving of Widgets as such with layout.walk

### DISCUSSION THREAD

**Discussion by CandussoR** [prompt-toolkit/python-prompt-toolkit#1929] [Ideas]
Hi everyone,

First time interacting with anyone on Github, I hope I do it the right way.

I've been investigating a little to see if I could use the lib to rework the way my console app flowed. 
I worked about ten hours with prompt_toolkit and the two things on which I passed most of my time was the layout and the flow of focus.

There are a lot of things we can see with introspection of the layout, but :
- It would be a lot easier if the UI elements, whether UIControls, Containers or Widgets would have a name property, like an ID in HTML. I tried to create custom elements, in the simplest way, to add a name property to them, but on introspection with `layout.walk()`, I only saw one type of element that I had created myself (the NamedHSplit). Therefore I couldn't grab the name of things like Frames or Windows directly, and still don't know how to interact them focus-wise.
- This bring me to the second point, that is : when walking the layout, it would be more readable if a Frame was a Frame, etc. We could just find what we need by using `isinstance()`, for example.

Maybe my problems just arise from the fact that I want a keyboard only app, and most developers and users design with mouse in mind ?

I thought I would share this with you, I don't know if anyone would find this interesting or not.

Have a good day!

**Comment by nitanmarcel** [comment]
Wouldn't `__name__` propriety work for you? Such as for example `NamedHSplit().__name__` will return `NamedHSplit` and you could create a method that recursively travels through the layout to get the the correct object

**Reply by nitanmarcel** [reply]
You could also inherit the things you need and add a name proprietary if that does it for you
